API: buffer notifications

KillTheMule opened the pull request:
Since @phodge seems to have run out of time, I hope you don't mind me taking over your PR. Thanks for your work so far :)

I've added an argument to `nvim_buf_live_udpates` that notes if the full buffer should be sent upon starting the updates. I've also added a test, should work.

TODO:

- [x] Adjust documentation
- [x] Rename functions after some bikeshedding
- [x] Replace `num_replaced` with `lastline`, akin to `set_lines`
- [x] Enlargen channel buffer capacity
- [x] Fix tests on windows
   - [x] Use ` new_pipename()` for the multiple channel test
   - [x] Don't fix the diffget test, but open an issue describing what happens, cc janlazo in it, see https://github.com/neovim/neovim/pull/7917#issuecomment-364788176 for a rough description
- [x] Add safeguards for the TODO conversion items. ~~Send a message to the client and abort (only happens with really large diffs/files on 32bit systems anyways).~~ Put in asserts for future proofness, can't happen today anyways.

Questions:

1. What should the functions be named? I figure `nvim_buf_live_updates` is fine, as long as we want to call this functionality `live updates`. Any thoughts?
2. What should the events be called? We have e.g. "LiveUpdateStart" and "LiveUpdate", which fits imho, but opinions are welcome.
3. ~~Is the documentation in the right place? Let me know where it should live. Is there another place this needs to be added to?~~
4. ~~As it stands now, registering and unregistering for the updates is done through the same function and   takes a boolean argument to distinguish between registering and unregistering, e.g. `nvim_buf_live_updates(true)` vs. `nvim_buf_live_updates(false)`. Since I've added a flag to maybe not send the whole buffer, those would be `nvim_buf_live_updates(true, true)` vs. `nvim_buf_live_updates(false, true)` where in the 2nd case the second argument is simply ignored. Feels ugly and weird. Should I make the 2 different functions? Or should there be a function to request the full buffer separately?~~
5. ~~As I mentioned in the [original PR](https://github.com/neovim/neovim/pull/5269), I'm getting trouble from the channel capacity. My idea would be to allow this to be configured by the plugin (i.e. introduce methods `nvim_get_channel_capacity` and `nvim_set_channel_capacity`), and keep the current default. I'd also think this should be another PR (I'd do it), so unless someone pipes up, I'll not include this here.~~
6. What else does this PR need in order to get merged?

All comments and ideas welcome, although please keep in mind I'm a beginner programmer. 

cc @justinmk, @bfredl, @phodge, @jamessan, @lunixbochs, @bryphe,@Chillee

Note: Unless you like function names like `function_to_make_nvim_send_lines_when_a_buffer_updated`, DON'T leave me alone with the naming issues. You have been warned :P

KillTheMule committed:
Update to latest master

KillTheMule committed:
Add argument to not send a buffers content when updates are enabled

Add a test.

KillTheMule committed:
Doc

bfredl commented:
>  We have e.g. "LiveUpdateStart" and "LiveUpdate", which fits imho, but opinions are welcome.

Events should have a prefix to distinguish builtin events from events sent by user code (`rpcnotify(..., "theevent")`). Unless there is a strong reason for CamelCase I think events should follow the general API convention, i e underscores and lower case. Also is "live" really good in the event name? All events about current state is "live"... It would better if the name signifies the actual change, that some lines of a buffer changes. Something in the roundabouts of `nvim_buf_lines_changed` I would expect.

KillTheMule committed:
Adjust FUNC_API_SINCE for nvim_buf_live_updates

KillTheMule commented:
Ok, how about

- `LiveUpdateStart` -> `nvim_buf_update_start`
- `LiveUpdate` -> `nvim_buf_update`
- `LiveUpdateTick` -> `nvim_buf_update_tick`
- `LiveUpdateEnd` -> `nvim_buf_update_end`

bfredl commented:
>  Or should there be a function to request the full buffer separately?

No, this would make it more easy to write racy code, subscription and full buffer send should be the same method call when applicable. I would rather have a separate method to stop subscription.

> My idea would be to allow this to be configured by the plugin (i.e. introduce methods nvim_get_channel_capacity and nvim_set_channel_capacity), and keep the current default. I'd also think this should be another PR (I'd do it), so unless someone pipes up, I'll not include this here.

I think we should attempt at reducing copying and allocation before we consider to add user configuration for memory limits.

KillTheMule commented:
> No, this would make it more easy to write racy code, subscription and full buffer send should be the same method call when applicable. I would rather have a separate method to stop subscription.

Ok, I will introduce a second function.

> I think we should attempt at reducing copying and allocation before we consider to add user configuration for memory limits.

I was afraid you were going to say something like this :D Are you positive a channel capacity of 10MB can be made sufficient to transfer a 1GB file in a sane number of notifications? Do you have something particular in mind?

bfredl commented:
I'm thinking of event names that maybe contains "lines" to mirror get_lines/set_lines. Also I would replace `numreplaced` with `last_line`, to follow the notion of `set_lines` (the numbers should be the same as a set_lines call that uses strict indexing, and nonnegative indicies)

bfredl commented:
How do you mean chunking solve the allocation problem? If the client hasn't read the first chunk when nvim wants to send the second, what happens?

lunixbochs commented:
I think it makes far more sense to block the channel when the buffer fills, encourage clients to drain the buffer as fast as they can (thus moving the buffer size explosion to the client), and have a user setting for client timeout, than it does for clients to ask neovim for more buffer size

KillTheMule commented:
> I'm thinking of event names that maybe contains "lines" to mirror get_lines/set_lines

We'd have to change that if we ever want to send partial lines though. That's why I'm kinda leaning towards not having `line` in the name. Otoh, if the return values change, a new name might not be the worst thing...

> How do you mean chunking solve the allocation problem? If the client hasn't read the first chunk when nvim wants to send the second, what happens?

I really don't know, I was hoping something along the lines of "chunk it up and hope the client fetches the chunks fast enough".



> Also I would replace numreplaced with last_line, to follow the notion of set_lines

Will do.

bfredl commented:
> I really don't know, I was hoping something along the lines of "chunk it up and hope the client fetches the chunks fast enough".

I'm sorry to be "that one" but "chunking" doesn't say me anything unless you specify what decision rule nvim will apply at the chunk boundary (that it couldn't apply otherwise)

KillTheMule commented:
> I'm sorry to be "that one" but "chunking" doesn't say me anything unless you specify what decision rule nvim will apply at the chunk boundary (that it couldn't apply otherwise)

No, worries, I'm really just "handwaving" in the worst possible way here. Note that the `LiveUpdateStart` notification already has a parameter `more` to notify that client that more lines are to be expected. That way, neovim can send several `LiveUpdateStart` notifications, each containing a set of lines so the message size is always <10MB.

bfredl commented:
Limiting chunk size surely has other benefits as it allows interleaving a smaller unrelated message without waiting for the full 1GB. But that is a separate problem (head-of-line problem).

I will investigate if memline can support COW-references without going insane (that would mean zero allocation of line _contents_ whenever the line is still in undo history, even if it was changed), until then (in this PR) I think we should just increase the memory limit.

KillTheMule commented:
> I will investigate if memline can support COW-references without going insane (that would mean zero allocation of line contents whenever the line is still in undo history, even if it was changed), until then (in this PR) I think we should just increase the memory limit.

Sounds good :) To what should I increase it? I went with 2GB locally, but I'm unsure if there are consequences of just using a large value.

bfredl commented:
Let's go for that unless someone protests. Alternatively, one possibility could be to make the memory quota shared among all channels, and at OOM close the one with largest memory consumption...

KillTheMule committed:
Make separate functions to start/stop live updates

bfredl commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -242,4 +242,198 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+`"nvim_buf_live_updates"`API method to activate Live Update events for a
+specific buffer. For example, in python >
+
+	import sys, neovim
```
I don't think this belongs here. At this point we can assume that the reader is aware of clients, attach-ment etc. Also we shouldn't start with a language-specific example, just describe the API functionality. (See `:help ui-intro` for comparison)

bfredl commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -242,4 +242,198 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+The co-process will start receiving the notification events which will be
+equivilent to the following |rpcnotify()| calls:
+
+1. rpcnotify({channel}, "LiveUpdateStart",		      *LiveUpdateStart*
```
`rpcnotfy` shouldn't be used here, the fundamental concept is msgpack notifications (and what's relevant for the reader is how they are represented at the receiving end, which we don't control). Just mention name + `[argument, list]`

KillTheMule replied:
Like [this](https://github.com/neovim/neovim/blob/master/runtime/doc/api.txt#L128)?

bfredl replied:
Possibly, but I would like to make events visually distinct from methods in some way, maybe by using `[ ]` argument list.

KillTheMule committed:
Lint

phodge committed:
API: Implement buffer updates

Originally written by @phodge in
https://github.com/neovim/neovim/pull/5269.

phodge committed:
API: Document buffer updates

Originally written by @phodge in
https://github.com/neovim/neovim/pull/5269.

phodge committed:
Tests for buffer updates

Originally written by @phodge in
https://github.com/neovim/neovim/pull/5269.

phodge commented:
@KillTheMule Thanks for picking this up! I haven't had energy to keep this going after so long, so it'd be absolutely awesome if you can get it across the finish line.

WRT to renaming events from "LiveUpdate" to "<something else>", remember to make the name something that is easy to google. :-)

WRT chunking ... when the LiveUpdateStart message is being generated, you could count the number of characters in each line as it is added to the array, and start a new chunk when the number of characters goes over some predefined limit. I don't know what the limit should be, but you could set it low enough that you never hit the channel buffer limits. And if there's a single line of code that's too many characters even on its own, you could just disable LiveUpdates and return an error message saying `Line X too many characters for live updates`. I think most users would be OK with having LiveUpdate-dependent plugins turn themselves off on the rare occasion they're editing a big file.

Thanks again for taking this on - I look forward to a future version of neovim with "Merged #7917" in the release notes. 👍

KillTheMule commented:
> WRT chunking ...

Yeah, that's basically what I meant to say :) But since the buffer capacity is 10MB right now, sending a 1GB file would probably produce a lot of overhead. We'll just go with upping the capacity for now.

@bfredl  So, for naming, we have the API functions, which are called `nvim_buf_live_updates_start` etc. right now, and the C functions with names like `liveupdate_register`. Should there be some similarity? I'll make an overview over the functions of this PR, so it's easier to think about :)

KillTheMule committed:
Make LiveUpdate return lastline instead of numreplaced

In analogy to `nvim_buf_set_lines`.

KillTheMule commented:
I made the changes @bfredl requested to the `LiveUpdate` event.

Here are the C functions we have (from `liveupdate.h`), and if applicable the corresponding API function, and the events sent:

- Functions:
  - `liveupdate_register` -> API `nvim_buf_live_updates_start`
  - `liveupdate_unregister` -> API `nvim_buf_live_updates_stop`
  - `liveupdate_unregister_all`
  - `liveupdate_send_changes`
  - `liveupdate_send_tick`
- Events:
  - `LiveUpdateStart`
  - `LiveUpdate`
  - `LiveUpdateTick`
  - `LiveUpdateEnd`

So how about we replace that with

- Functions:
  - `buffer_updates_register` -> API `nvim_buf_updates_start`
  - `buffer_updates_unregister` -> API `nvim_buf_updates_stop`
  - `buffer_updates_unregister_all`
  - `buffer_updates_send_changes`
  - `buffer_updates_send_tick`
- Events:
  - `nvim_buf_updates_start`
  - `nvim_buf_update`
  - `nvim_buf_update_tick`
  - `nvim_buf_updates_end`

Just throwing out ideas, get me something better (or what you like better, I don't mind), and I'll put it in :)

KillTheMule commented:
The filter test fails on windows because we don't have `:!sort`. Is there something else I can use, or should I skip that test on windows?

Also, [this](https://github.com/KillTheMule/neovim/blob/aa146e7b95f4f9c49b685803143bb8790f0f90c5/test/functional/api/liveupdate_spec.lua#L547) fails on windows with `Vim(diffthis):E97: Cannot create diffs`, does anyone know what's up with this?

Third, [this](https://github.com/KillTheMule/neovim/blob/5a3779875732ceb425ee27882f90ecc61a282fed/test/functional/api/liveupdate_spec.lua#L466) fails on windows with `test\functional\helpers.lua:93: Failed to evaluate expression`, does anyone know those things.

@janlazo You seem to know a lot about windows things, could you have a look? Unfortunately, I can't test on windows locally...

KillTheMule committed:
Bump up buffer capacity to 2GB

KillTheMule committed:
Fix tests on windows

`cat` is distributed with neovim, so when can use it everywhere, as
opposed to `sort`.

The diffget test fails for unknown reasons on appveyor, mark it pending
for now.

KillTheMule committed:
Fix memory leak

bfredl commented on test/functional/api/liveupdate_spec.lua:
```diff
@@ -0,0 +1,749 @@
+      where = alter_slashes(where)
+      print("serverstart('"..where.."')")
+      --eval("serverstart('"..where.."')")
+      helpers.command("call serverstart('"..where.."')")
```
Do we really need multiple servers for this? Can't we just connect multiple times to `v:servername` ?

KillTheMule replied:
Yep, adjusted that.

KillTheMule commented:
QB/Travis failures unrelated. AV failure seems geniune, but I have not idea how to go about it, I could not reproduce it manually. I'll do some desperate debugging attempts, but hopefully someone can lend me a hand.

Other than that, I think what's needed is a decision about naming. That's it :)

KillTheMule commented:
I'm giving up on the diffget test on windows, I'll open an issue describing my findings, and I'll mark it pending.

Notes to self

- File `tmp_diff` in the C stays empty
- Diff tool on appveyor returns 0 (that is, the C function call that calls the shell that calls... returns 0, which should be the return value of the diff tool... it's ignored though), while when locally testing on linux it returns 1
- Tried `:let &shcf='/s /c'` and `:let &sxq='"'` per justinmks suggestion to no avail.

KillTheMule committed:
The grand renaming

KillTheMule committed:
Enable -Wconversion

KillTheMule commented:
I went with the naming I suggested in https://github.com/neovim/neovim/pull/7917#issuecomment-360987656. I'm preparing to finish this up, so if anyone's interested in the naming, please speak up soon.

janlazo commented:
How does `:diff` escape the filenames for `diff.exe`? If the problem is `shellescape`, then setting `diffexpr` with a custom shellescape (https://github.com/neovim/neovim/pull/7786/commits/ca5f2d73be570666257dcc666164be19fffff45d) for each filename, including the tempfile, should fix it. `set shellcmdflag=/s/c shellxquote=\" is still required. That shellescape is not bullet-proof because of cmd.exe quirks (I have my own version to handle the quirks) but it should be enough for diff.exe

KillTheMule commented:
Thanks @janlazo for your comment, I will however keep the test pending and make an issue, I don't want to get distracted too much on this PR. I'd have to adapt your custom shellescape to C and use that for every diff call, that sounds too much for something that as-of-now only happens on appveyor. I'll be sure to copy that into the issue I'll open though :)

In other news, I've enabled `-Wconversion` for the new code, but some casts are needed that could theoretically be wrong. After discussion on gitter with justinmk, bfredl and jamessan I've opted for some `STATIC_ASSERT`s  which should not fail on any plattform we support now, but might on a future plattform, so the casts can be revisited then.

Other than that, I think this is done. Let's wait what CI has to say.

KillTheMule committed:
Increase sendkeys timeout

KillTheMule committed:
Try fixing that test on travis

KillTheMule commented:
All build failures are unrelated: Travis lint timed out, Appveyor 32bit timer_spec, appveyor 64bit gcov the usual, QB MacOs job_spec, QB Linux kbtree stuff.

So I'm marking this RDY. Let me know if there's something else to do, or if I should squash commits more (will probably squash the last and remove the penultimate one).

bfredl commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -242,4 +242,170 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+                                                              *buffer-updates-api*
+BufferUpdates Functions~
+
+nvim_buf_updates_start({send_buffer})                   *nvim_buf_updates_start()*
```
Function docs with argument lists will be auto-generated from the doxygen docs, so these shouldn't be duplicated here. See `:help api-buffer` for example of the results. (Which means it's also fine for the docs inside `api/*.c` to use helptags, so the reference to the events can be moved there)

KillTheMule replied:
I'm being a bit dense, sorry. I've looked at `:h api-buffer`, and it looks fine. But that document does not contain `nvim_buf_updates_start`, do I need to do something special to get it in there? I'll scratch those function definitions, but please let me know what I need to do for the api docs to work.

bfredl replied:
They will be generated before release at latest, so don't worry.

bfredl replied:
But `scripts/gen_api_vimdoc.py` if you're curious

KillTheMule replied:
Ahh now that's comfortable :)

bfredl commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -242,4 +242,170 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+If buffer updates are activated a new empty buffer, the following
+|nvim_buf_updates_start| event will be sent: >
+
+  rpcnotify({channel}, "nvim_buf_updates_start", [{buf}, [""], v:false])
```
also don't use `rpcnotify` here.

KillTheMule replied:
Right, I removed that and just gave the events as examples.

Chillee commented:
What's the current progress of this PR?

KillTheMule commented:
From my PoV, this is done, I'm waiting for a review. As far as I can tell from gitter, the devs might be waiting for feedback from someone trying to use that other than me.

justinmk commented on src/nvim/buffer_updates.c:
```diff
@@ -0,0 +1,214 @@
+// Register a channel. Return True if the channel was added, or already added.
+// Return False if the channel couldn't be added because the buffer is
+// unloaded.
+bool buffer_updates_register(buf_T *buf, uint64_t channel_id, bool send_buffer)
```
file names are buffer.c/buffer_updates.c but all functions always start with `buf_`, not `buffer_`. Let's stay consistent on this.

justinmk commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -242,4 +242,150 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+
+nvim_buf_updates_start[{buf}, {changedtick}, {linedata}, {more}]	 *nvim_buf_updates_start*
+
+  Neovim will send at least one of these notifications to confirm that
```
Neovim is spelled "Nvim" in the docs, or "nvim" when referring to the command.

justinmk commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -242,4 +242,150 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+
+nvim_buf_update_tick[{buf}, {changedtick}]		                   *nvim_buf_update_tick*
+
+  Indicates that |b:changedtick| was incremented for the buffer {buf}, but no
```
If this event is defined by `b:changedtick`, it should use the same jargon in the event name: `nvim_buf_changedtick` ?

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -3150,7 +3165,8 @@ static char_u *sub_parse_flags(char_u *cmd, subflags_T *subflags,
 /// @return buffer used for 'inccommand' preview
-static buf_T *do_sub(exarg_T *eap, proftime_T timeout)
+static buf_T *do_sub(exarg_T *eap, proftime_T timeout,
+                     bool send_buffer_update_changedtick)
```
avoid overly verbose names. suggestion: `do_buf_event`

justinmk commented on src/nvim/undo.c:
```diff
@@ -1732,7 +1735,7 @@ bool u_undo_and_forget(int count)
 
 /// Undo or redo, depending on `undo_undoes`, `count` times.
-static void u_doit(int startcount, bool quiet)
+static void u_doit(int startcount, bool quiet, bool send_update)
```
`do_buf_event` + `@param` docstring

justinmk commented:
Are we still considering [back pressure](https://github.com/neovim/neovim/pull/7917#issuecomment-360828124) as a TODO?

KillTheMule commented:
> Are we still considering back pressure as a TODO?

I never did, because for one it was only one person vouching for it, and I can't personally judge if that's really the thing to do. Also, it would be over my head to implement this.

> Or is "chunking" preferred?

That has the drawback that one might end up with a lot of chunks. Sending that 2GB buffer over a 10MB sounds unreasonable to me.

KillTheMule committed:
Some renamings and doc changes

KillTheMule committed:
Update test

KillTheMule committed:
Lint

KillTheMule commented:
I made the changes you asked for and fixed the test. Let me know if I should squash more, I'd guess that changing older commits during a review is less than helpfull, but if I'm wrong and I can make it easier for you by squashing, no problem :)

justinmk commented on test/functional/api/buffer_updates_spec.lua:
```diff
@@ -0,0 +1,737 @@
+  return tick
+end
+
+describe('liveupdate', function()
```
describe('buffer events', function()

and "knows" can be removed from the descriptions below.

KillTheMule commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -255,11 +255,11 @@ The co-process will start receiving the following notification events:
-  contents. If the buffer is very large, neovim might send the contents through
-  in multiple events to avoid loading the entire buffer's contents into
-  memory at once.
+  Nvim will send at least one of these notifications to confirm that buffer
```
What the... darn, smartcase at work :D

bfredl commented:
Currently the start request has the same name as the first event which is a bit unfortunate. On gitter @justinmk and I discussed using event subscription as the language for these kind of requests, so something like  `nvim_buf_event_sub` for the method.

KillTheMule commented:
Ok, sooo... a function `nvim_buf_event_sub` needs arguments. Firstly, an event. That would just be a string, right? But how to handle args, different event subscriptions might need different args. We'd also need `nvim_buf_event_unsub`, right?

bfredl commented:
My understanding was to just rename the current method to `nvim_buf_event_sub` or something like that.
An options dict can and probably should be added for further extensions, but using an options dict to squeeze _different_ requests (with different possible args) into the same formal API method should be considered an anti-pattern.

KillTheMule commented:
> An options dict can and probably should be added for further extensions, but using an options dict to squeeze different requests (with different possible args) into the same formal API method should be considered an anti-pattern.

Ok, I'll just rename it, but since you mention it... isn't this just what's being done with `nvim_call_function`?

bfredl commented:
> Ok, I'll just rename it, but since you mention it... isn't this just what's being done with nvim_call_function?

Not sure I follow. The point is we shouldn't deliberately introduce an API function that corresponds to a _limited_ set of logical API requests, just because they happen to have a stream of events instead of a single sync return value. From that it doesn't follow we can't have _generic_ FFI functionality in the API (that BTW also includes _user-defined_ functions and thus never could have been covered by a finite set of API functions in the first place), no?

justinmk commented:
Quickbuild failures are unrelated.

Thanks everyone!

justinmk commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -241,4 +241,152 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+
+  {more} is a boolean which tells you whether or not to expect more
+  |nvim_buf_updates_start| notifications. When {more} is false, you can be certain
+  that you now have the entire buffer's contents.
```
Why don't we use this same mechanism for `nvim_buf_update`? And why is a separate `nvim_buf_updates_start` event needed?

KillTheMule replied:
Not needed. I agree it would be nicer to just send `nvim_buf_update` for everything, and have the `{more}` parameter also for non-initial updates, even if we don't use it yet. #workingonit

KillTheMule replied:
As a user of this feature, I'm not particularly thrille having to specialcase for `lastline=-1`. Otoh, I specialised for the event before, so it's not a huge deal.

Just to be clear, if a non-initial sending includes the last line or even the whole buffer, we still send "real" numbers, not -1, right?

bfredl replied:
Well, previously you had to specialcase an entire separate event.

> Just to be clear, if a non-initial sending includes the last line or even the whole buffer, we still send "real" numbers, not -1, right?

Yes. In that case you are still replacing a range of lines that the client already knows about, which could happen to be the previous entire buffer, so this should still use the concrete length. But with the first send there is no previous content for the client to replace (it knows nothing, or rather nvim doesn't know if it knows anything ), so specifying a concrete length doesn't make sense. Rather `[0, -1]` would mean "regardless of previous state of information (probably nothing), these lines are the entire buffer".

justinmk replied:
@KillTheMule fair point, but clients already would have to handle `nvim_buf_updates_start` as a special case.

The semantics of `-1` are well established by e.g `nvim_buf_{get,set}_lines`.

KillTheMule replied:
Actually, do we need a `{more}` parameter? You always need to expect ~spanish~ buffer updates. So if you request the buffer's content, and get something that's not the whole buffer, the next update will get you up-to-date, the line numbers will tell you specifics.

Right?

That of course means that we can't use `-1` for the initial send (at least if it's not the whole buffer), but I don't think that's really needed for anything.

bfredl replied:
> That of course means that we can't use -1 for the initial send (at least if it's not the whole buffer), but I don't think that's really needed for anything.

I don't think that follows. The first send still starts from an unknown state, but the following (even if they help build the "true" initial state) builds from the incomplete but concrete client state.

But I still think we want to somehow separate "real" buffer states, from ones invented for transport technical reasons. How about setting changedtick to `-1` in the cases that would have `more`, if we want to avoid this flag? Because these do not correspond to any real buffer state, while true buffer states are identified by their changedtick.

bfredl replied:
>  One can identify those because the changedtick stays the same, if need be, but I don't really see why that would be neccessary.

No you cannot, as you when you receive one event don't yet know if more events will come with the same changetick. But if separate `more` flag (which I imagined would be usable _beyond_ the initial send, for similarly large updates) is preferred, let's use that :)

justinmk replied:
> which I imagined would be usable beyond the initial send, for similarly large updates

That's what I was thinking, seems like `more` ties in with the "chunking" discussion near the top of this PR.

justinmk commented on src/nvim/api/buffer.c:
```diff
@@ -74,6 +75,50 @@ String buffer_get_line(Buffer buffer, Integer index, Error *err)
+/// @param[out] err Details of an error that may have occurred
+/// @return False when updates couldn't be enabled because the buffer isn't
+///         loaded; otherwise True.
+Boolean nvim_buf_event_sub(uint64_t channel_id,
```
rename to `nvim_buf_attach`

justinmk commented on src/nvim/api/buffer.c:
```diff
@@ -74,6 +75,50 @@ String buffer_get_line(Buffer buffer, Integer index, Error *err)
+/// @param[out] err Details of an error that may have occurred
+/// @return False when updates couldn't be disabled because the buffer
+///         isn't loaded; otherwise True.
+Boolean nvim_buf_event_unsub(uint64_t channel_id,
```
rename to `nvim_buf_detach`

justinmk commented on src/nvim/buffer_updates.h:
```diff
@@ -0,0 +1,16 @@
+                              int64_t num_added,
+                              int64_t num_removed,
+                              bool send_tick);
+void buf_updates_changedtick(buf_T *buf);
```
why are these declarations needed in the header file? Normally our build system does this automatically. Try this instead:

```c
#ifdef INCLUDE_GENERATED_DECLARATIONS
# include "buffer_updates.h.generated.h"
#endif
```

justinmk commented:
Quickbuild failures are unrelated.

@KillTheMule As we discussed in chat, I think we should rename the concept to "attach". This was even suggested in #2224. 

---
closes #2224
closes #5269
closes #1114

KillTheMule committed:
Use autogenerated declarations

KillTheMule committed:
Unify updates_start and updates to lines_event

Also rename changedtick -> changedtick_event

KillTheMule commented:
Ok, I implemented everything we talked about, adjusted the tests and docs.

KillTheMule commented:
Ok, done. If you don't request an initial buffer sending, the event will have arguments `firstline=-1, lastline=-1`, as opposed to requesting the sending on an empty buffer, which will result in `firstline=0, lastline=-1` (as will all requested initial sendings). Is that right? Both will not have lines attached to them, and I figured being able to distinguish them via `firstline` isn't bad.

KillTheMule committed:
Send changedtick as first event if buffer contents weren't requested

KillTheMule commented:
CI failures are unrelated (Travis has that spell test failing, QB has icm + term, jobspec + 2x timer-spec).

KillTheMule commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -241,4 +241,124 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+If the user adds 2 new lines to the start of a buffer, the following event
+would be generated: >
+
+  nvim_buf_update[{buf}, 0, 0, ["line1", "line2"]]
```
Sorry I missed that!

KillTheMule commented on runtime/doc/msgpack_rpc.txt:
```diff
@@ -241,4 +241,124 @@ Even for statically compiled clients it is good practice to avoid hardcoding
+
+    {changedtick} is the new value of |b:changedtick| for that buffer.
+
+nvim_buf_updates_end[{buf}]                                    		*nvim_buf_updates_end*
```
Will do!

KillTheMule committed:
Rename some more, fixe borked renaming

KillTheMule commented:
Notes to self: Let `attach` take an unused options dict, cf `nvim_buf_get_commands`. Document that if you don't request initial buffer contents you get a `changedtick_event`.

bfredl commented:
I think receiving initial state (total state of information, no risk of incorrect synchronization) should be the default.

justinmk commented:
> I think receiving initial state (total state of information, no risk of incorrect synchronization) should be the default.

Though as discussed on gitter, the nodejs https://www.npmjs.com/package/msgpack-lite lib seems very slow, so it's good that it's optional.

bfredl commented:
@justinmk but it's gonna have the same issue when the user does a large edit,like a global substitution or reindent the file.

KillTheMule commented:
> I think receiving initial state (total state of information, no risk of incorrect synchronization) should be the default.

There's no default though since the parameter isn't optional. You send `true`, you get the buffer, you send `false`, you don't. Or am I misunderstanding you?

bfredl commented:
@KillTheMule your comment I replied to as well as gitter discussion at the time made it sound we planned to include that option in the option dict. If we don't do that (which is probably better, as it is quite an important option), the point is moot of course.

KillTheMule committed:
Add empty options dict to buf_attach

KillTheMule committed:
API: Accept empty lists as dictionaries

KillTheMule commented:
As per gitter discussion, I've added a commit that makes the api accept an empty array as an empty dict. Please review carefully, and maybe suggest a test. In particular, the cast in the second half of https://github.com/neovim/neovim/pull/7917/commits/4fb003330575ba4f010743a84bb22817f929783c#diff-49dac2d54175bc1b0f98105128d37766R228 seems like "Screw, meet Hammer!" to me, is there a better way?

If you want, I can make a separate PR for that, of course.

KillTheMule commented:
CI failures seem unrelated: Travis Clang has an apt-get failure, travis xcode has that msgpack failure, QB macosx has failures in timer_spec.lua, QB linux release on inccomand_spec.lua. Appveyor gcov is known I guess, and there's a failure in [`job_spec.lua`](https://ci.appveyor.com/project/neovim/neovim/build/20585/job/1tj6ue56insvccus#L10322) on msvc32, which passes on all other builds. I'll rebase and push to see if that fails again.

KillTheMule commented:
Ok, appveyor passed without any change, so it was just flaky. The new failures on the other CIs are: Travis linux gcc 32bit failure on ci/before_script.sh, travis osx clang failure in spell_spec.lua, travis osx gcc msgpack-rpc invalid string, QB osx debug failure in timer_spec.lua, QB osx release failure in job_spec.lua and timer_spec.lua, QB freebsd release no idead (all tests pass, still a negative result).

Quite a bit, but as far as I saw, most of those are really occuring more often than not. If you think any of those need closer investigation for this PR, please let me know.

ZyX-I commented:
Working around lua bindings problems on the side of Neovim seems wrong. If argument type is a dictionary, it should receive a dictionary. Logical way to fix the problem is modifying lua client, you may see me frequently using “typed tables” in unit tests for creating VAR_DICT or Dictionary values. Here by “typed tables” I mean “tables which have a key that cannot possibly be send as a Dictionary key with attached value specifying a type”. E.g. see https://github.com/ZyX-I/neovim/blob/42f2327ac741ffa9d757040fdecc75069b4d3827/test/unit/eval/helpers.lua#L371.

ZyX-I commented:
This concept is also useful for forcing integral value to be send as a float. And specifically there I used it to create VAR_FUNC and VAR_PARTIAL as well, though this is not applicable for API types (yet).

bfredl commented:
@ZyX-I something like that would be needed to pass empty dict as Object, but it seems overblown to need sentinel key just to pass an empty mandatory options dict. I would expect `curbufmeths.attach(true, {})` to work, this clearly expresses the intent, which is calling the api function with no options, not constructing a dict for the sake of it. Already 0/1 can be used for boolean args in languages that don't cleanly separate bools and ints (while still either native or replacement bool will be needed for a Object value), and also int instead of typed handle.

bfredl commented:
Well, lua is a language we have chosen, I don't see the point of arguing it is somehow a rare edge case for us.

> This makes API be loosely typed for no reason other then minor convenience of a set of languages, which is also undetermined as far as I know. If language has enough capabilities for creating binary string like msgpack it surely has enough capabilities for creating correctly typed string, even though it may increase some code lines lengths. 

The API is defined in terms of its own type system. Msgpack is something we _use_ to represent the API on the wire, it is not what fundamentally defines the meaning of API types. The essence of Buffer type is that it a handle that represents a buffer, not that it is represented by a set of msgpack values disjoint from any other concrete type. What msgpack strings are "correct" for a top-level type is something that we decide.

>  I would rather purge out all existing conversions and make API strictly typed then add any more

Being able to call an buffer method on the current buffer or known buffer number, by passing msgpack int is a simple and useful _feature_, an API consumer shouldn't need to construct a msgpack special value just to get started and call methods on the current buffer. A set of typing rules that would rather purge such useful features, would care about the wrong things and are not the rules I think we should follow. Similarly for lua, I wouldn't expect sentinel value to be needed just to call API function with dict for optional parameters, for the case with no options. If options dict is a pattern we want to use with API functions, this will become a common case, not a rare case.

bfredl commented:
@chemzqm In general attachment is on a channel basis (same as UI). Rather the host should allow multiple plugins to share a buffer attachment. I might prototype this in the python host, to show the general pattern. This will have other benefits: in any sync context `buf[line:range]` (python syntax, but the idea should work the same) can use the cached value in the host instead of needing to send an additional request.

chemzqm commented:
@bfredl I've tested that this could happen for different node plugins
``` js
let buffer = null

async function getLines () {
  let {nvim} = this
  buffer = await nvim.buffer
  await buffer.request('nvim_buf_attach', [buffer, false, {}])
}

async function unbind() {
  await buffer.request('nvim_buf_detach', [buffer])
}

module.exports = (plugin) => {
  plugin.registerCommand('LiveListen', getLines.bind(plugin), {sync: false})
  plugin.registerCommand('LiveunListen', unbind.bind(plugin), {sync: false})
  plugin.nvim.on('notification', (name, args) => {
    if (name === 'nvim_buf_lines_event') {
      let [buf, tick, firstline, lastline, linedata, more] = args
      console.error(`changes: ${firstline} ${lastline}`)
    }
  })
}
```
``` js
let buffer = null

async function getLines () {
  let {nvim} = this
  buffer = await nvim.buffer
  await buffer.request('nvim_buf_attach', [buffer, false, {}])
}

module.exports = (plugin) => {
  plugin.registerCommand('LiveNotify', getLines.bind(plugin), {sync: false})
  plugin.nvim.on('notification', (name, args) => {
    if (name === 'nvim_buf_lines_event') {
      let [buf, tick, firstline, lastline, linedata, more] = args
      console.error(`notify: ${firstline} ${lastline}`)
    }
  })
}
```
Detach called in one plugin would cause notification not working in other plugins.

There is a detach event `nvim_buf_detach_event`, but how could I attach event listener to the buffer again if it was triggered by other plugin calling `detach`?

bfredl commented:
> There is a detach event nvim_buf_detach_event, but how could I attach event listener to the buffer again if it was triggered by other plugin calling detach?

Well the point is that the plugin won't call the nvim methods directly, rather helper functions in the host. The host can keep refcounting on the buffer, and only call detach when the number of plugins using it goes to zero. I think I have an old unfinished branch for this in python, seems like a good time for me to revive it....

chemzqm commented:
@bfredl But there would some other problem for a host to manage the count of `attach` and `detach`.    
For example, a plugin may call `attach` on different user event(the attach could be called more than one time), and it would expect one `detach` call to detach the events.
So it requires something like plugin id to make the `detach` work properly when one plugin call `attach` multiply times.

bfredl commented:
@chemzqm That sounds to me more like an "accidental" problem (change the helper functions to expose the most convenient interface to plugins) rather than a "fundamental" problem (that helper function solution won't work at all)

My thought for python host would be to just reify the subscription rather than introduce plugin identity, so something like

`id = thebuf.attach(self.callback, options...)`

and the plugin could later do `thebuf.detach(id)` or alternative `attach` would return an object with a `.detach()`. So detach is called the same number times as attach (or not at all, if subscription is persistent) Won't something like this make sense for nodejs as well?

___

The alternative for me, if we want nvim core to be more involved, would be to introduce some kind of sub-channel routing mechanism, where each `nvim_..._event` will have an adress handle (or list of such) as first parameter. (but even then I think host will be involved, because we will want to multicast the some updates to different addresses on a channel, to not duplicate all buffer updates). To me this seems overblown for this PR as it stands, but maybe this will be useful with further expansions, like individual plugin only caring about extmarks in its own namespace.

KillTheMule commented:
Wait, am I missing something? Registering for updats happens via channel_id. Am I wrong in thinking that every remote plugin gets its own channel? So what I think you're describing shouldn't happen, @chemzqm . I can't read your code, but maybe I can cook up a test for that.

Or am I off with something here?

justinmk commented:
Nicely done. Thank you @phodge @KillTheMule and @bfredl .

justinmk commented on src/nvim/terminal.c:
```diff
@@ -1234,7 +1234,8 @@ static void refresh_screen(Terminal *term, buf_T *buf)
   int change_end = change_start + changed;
-  changed_lines(change_start, 0, change_end, added);
+  // Note: don't send nvim_buf_lines_event event for a :terminal buffer
+  changed_lines(change_start, 0, change_end, added, false);
```
@KillTheMule is there at test to check that `:terminal` buffers do not trigger `nvim_buf_lines_event`?

KillTheMule replied:
No there isn't. But now that you mention it: Why shouldn't they? I mean, they're buffers after all, if someone wants to subscribe to their updates, I don't see I reason why we should deny them. Or is there a specific reason I'm not seeing?

justinmk replied:
Sure, why not. I was just going by the comment here.

Would still want a test either way

KillTheMule replied:
I'll send a PR, might take some time though.

chemzqm replied:
I want to monitor the content of a terminal buffer that loaded from session file,  it seems no other options except using the buffer updates API.

KillTheMule commented:
No there isn't. But now that you mention it: Why shouldn't they? I mean, they're buffers after all, if someone wants to subscribe to their updates, I don't see I reason why we should deny them. Or is there a specific reason I'm not seeing??

bfredl commented:
I began to look into python host integration: https://github.com/neovim/python-client/pull/340

justinmk commented on src/nvim/generators/gen_api_dispatch.lua:
```diff
@@ -223,6 +223,11 @@ for i = 1, #functions do
           output:write('\n  } else if (args.items['..(j - 1)..'].type == kObjectTypeInteger && args.items['..(j - 1)..'].data.integer >= 0) {')
           output:write('\n    '..converted..' = (handle_T)args.items['..(j - 1)..'].data.integer;')
         end
+        -- accept empty lua tables as empty dictionarys
```
"lua" in the above comment is a bit misleading. This applies to all msgpack requests--though it's probably only _useful_ for a Lua client.

KillTheMule replied:
Right, I was somewhat in "lua-land" when I did this.
